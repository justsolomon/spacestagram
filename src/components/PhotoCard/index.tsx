import { Photo } from "types/photo";
import LikeButton from "components/LikeButton";
import styles from "./photoCard.module.scss";
import { useInView } from "react-intersection-observer";
import ShareMenu from "components/ShareMenu";
import PlaceholderImage from "assets/images/placeholder.jpeg";
import usePhotoStorage from "hooks/usePhotoStorage";

interface PhotoCardProps {
  /** Mars Rover photo information */
  photo: Photo;
}

const PhotoCard = ({ photo }: PhotoCardProps) => {
  const { ref, inView } = useInView({ threshold: 0.25, triggerOnce: true });
  const { isLiked, toggleLike } = usePhotoStorage(photo.id);

  const title = `${photo.rover.name} rover - ${photo.camera.full_name}`;

  return (
    <li className={styles["photo-card"]} ref={ref} data-testid="photo-card">
      <img
        src={inView ? photo.img_src : PlaceholderImage}
        alt={title}
        loading="lazy"
        className={styles["photo-card__image"]}
      />
      <div className={styles["photo-card__section"]}>
        <div className={styles["photo-card__details"]}>
          <p className={styles["photo-card__title"]}>{title}</p>
          <p className={styles["photo-card__date"]}>{photo.earth_date}</p>
        </div>
        <div className={styles["photo-card__action-buttons"]}>
          <ShareMenu
            rover={photo.rover.name}
            camera={photo.camera.full_name}
            imageSrc={photo.img_src}
          />
          <LikeButton liked={isLiked} title={title} toggleLike={toggleLike} />
        </div>
      </div>
    </li>
  );
};

export default PhotoCard;
