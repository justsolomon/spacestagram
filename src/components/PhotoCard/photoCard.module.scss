@import "styles/variables";

.photo-card {
  width: 100%;
  border-radius: $border-radius--sm;
  border: 1px solid $bg-dark--primary;
  background-color: $white;

  &__image {
    width: 100%;
    height: 200px;
    border-top-left-radius: $border-radius--sm;
    border-top-right-radius: $border-radius--sm;
    background-color: $bg-light--tertiary;
    object-fit: cover;
  }

  &__section {
    padding: 0.8rem 1rem;
    display: flex;
    justify-content: space-between;
    align-items: center;
  }

  &__details {
    width: 71%;
    line-height: $font-size--lg;
  }

  &__action-buttons {
    width: 28%;
    display: flex;
    justify-content: flex-end;
  }

  &__title {
    font-weight: $font-weight--medium;
    font-size: $font-size--md;
    line-height: $font-size--xl;
  }

  &__date {
    color: $secondary-color;
    margin-top: 0.3rem;
  }

  @media #{$media-tablet} {
    &__details {
      width: 78%;
    }

    &__action-buttons {
      width: 21%;
    }
  }

  @media #{$media-mobile} {
    &__details {
      width: 75%;
    }

    &__action-buttons {
      width: 24%;
    }
  }

  @media #{$media-mobile-sm} {
    &__section {
      padding: 0.5rem 0.8rem;
    }

    &__details {
      width: 73%;
      font-size: $font-size--sm;
      line-height: $font-size--md;
    }

    &__title {
      font-size: $font-size--base;
    }

    &__action-buttons {
      width: 26%;

      button {
        margin-left: 0.05rem;
      }
    }
  }
}
