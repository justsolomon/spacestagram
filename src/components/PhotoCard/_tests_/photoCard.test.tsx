import { render, screen } from "@testing-library/react";
import { act } from "react-dom/test-utils";
import PhotoCard from "..";
import { intersectionMockInstance } from "react-intersection-observer/test-utils";

const photo = {
  id: 424926,
  sol: 1000,
  camera: { id: 22, name: "MAST", rover_id: 5, full_name: "Mast Camera" },
  img_src:
    "http://mars.jpl.nasa.gov/msl-raw-images/msss/01000/mcam/1000ML0044631200305217E01_DXXX.jpg",
  earth_date: "2015-05-30",
  rover: {
    id: 5,
    name: "Curiosity",
    landing_date: "2012-08-06",
    launch_date: "2011-11-26",
    status: "active",
  },
};

describe("<PhotoCard />", () => {
  it("should render the photo card correctly", () => {
    const { asFragment } = render(<PhotoCard photo={photo} />);

    //mock intersection observer
    act(() => {
      intersectionMockInstance(screen.getByTestId("photo-card"));
    });

    expect(asFragment()).toMatchSnapshot();
  });
});
