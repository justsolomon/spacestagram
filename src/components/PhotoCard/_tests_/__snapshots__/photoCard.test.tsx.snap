// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PhotoCard /> should render the photo card correctly 1`] = `
<DocumentFragment>
  <li
    class="photo-card"
    data-testid="photo-card"
  >
    <img
      alt="Curiosity rover - Mast Camera"
      class="photo-card__image"
      loading="lazy"
      src="placeholder.jpeg"
    />
    <div
      class="photo-card__section"
    >
      <div
        class="photo-card__details"
      >
        <p
          class="photo-card__title"
        >
          Curiosity rover - Mast Camera
        </p>
        <p
          class="photo-card__date"
        >
          2015-05-30
        </p>
      </div>
      <div
        class="photo-card__action-buttons"
      >
        <button
          aria-label="Like Curiosity rover - Mast Camera photo"
          aria-pressed="false"
          class="like-button "
          data-testid="like-button"
        >
          <svg>
            heart-outline.svg
          </svg>
        </button>
      </div>
    </div>
  </li>
</DocumentFragment>
`;
